### PR TITLE
Extract and store author name from FxOS Add-ons manifests (bug 1209582)

### PR DESCRIPTION
**Pull request by diox**
https://bugzilla.mozilla.org/show_bug.cgi?id=1209582

Bonus: add searching by author to the API.

### CODE REVIEW

File: mkt/extensions/validation.py [hunk 1]
```diff
@@ -168,3 +171,12 @@ def validate_version(self, manifest_json):
             except ValueError:
                 # Not a valid number.
                 raise ParseError(self.errors['VERSION_INVALID'])
+
+    def validate_author(self, manifest_json):
+        """
+        Ensures that, if present, the author property is no longer than
+        255 characters.
+        """
+        description = manifest_json.get('author')
```
Comment (eviljeff) [inline mkt/extensions/validation.py]: description?

File: mkt/extensions/validation.py [hunk 2]
```diff
@@ -19,6 +19,8 @@ class ExtensionValidator(object):
     error.
     """
     errors = {
+        'AUTHOR_TOO_LONG': _(
+            u'The `author` property cannot be longer than 255 characters.'),
```
Comment (eviljeff) [inline mkt/extensions/validation.py]: Device side UX would prefer 128 and that seems enough to me.  Can we change to 128?

File: mkt/extensions/validation.py (hunk 1 above)
Comment (diox) [inline mkt/extensions/validation.py]: oops

File: mkt/extensions/validation.py (hunk 2 above)
Comment (diox) [inline mkt/extensions/validation.py]: Sure. Do you have a link for the decision ? I was looking for one and couldn't find any.

Comment (eviljeff) [inline mkt/extensions/validation.py]: I just asked Jacqueline on IRC.  And then confirmed with David Almstrom. If there is a wiki page for the spec it should be (added) there though

Comment (chuckharmston) [inline mkt/extensions/validation.py]: https://wiki.mozilla.org/Marketplace/Add-ons/Validation

Comment (diox) [inline mkt/extensions/validation.py]: Updated to 128. A little weird since description is at 132 though, but I guess that's fine ? I also added a bunch of validation tests.

File: docs/api/topics/firefox_os_addons.rst [hunk 3]
```diff
@@ -133,12 +135,15 @@ Search
...
-    * If a query is passed, order by relevance.
-    * If no query is passed, order by popularity descending.
+    All query parameters are optional. The default sort order when the `sort`
+    parameter is absent depends on whether a search query (`q`) is present or
+    not:
+    * If a search query is passed, order by relevance.
+    * If no search query is passed, order by popularity descending.
 
     :param string q: The search query.
+    :param string author: Filter by author. Requires a case-insensitive
```
Comment (chuckharmston) [inline docs/api/topics/firefox_os_addons.rst]: `string|null`?

File: docs/api/topics/firefox_os_addons.rst [hunk 4]
```diff
@@ -68,6 +69,7 @@ Detail
         depending on the query. See :ref:`translations <overview-translations>`.
 
     :resjson int id: The add-on id.
+    :resjson string author: The add-on author, if specified in the manifest.
```
Comment (chuckharmston) [inline docs/api/topics/firefox_os_addons.rst]: `string|null`?

File: mkt/extensions/indexers.py [hunk 5]
```diff
@@ -16,12 +16,17 @@ class ExtensionIndexer(BaseIndexer):
...
+        'trending_*',
+        'boost',
+        # 'name' and 'description', as well as their locale variants
+        # ('name_l10n_<language>', etc.) are only used for the query matches,
+        # and are never returned to the client through the API. The fields that
+        # are returned to the API are '*_translations'.
         'name',
         'name_l10n_*',
+        'description',
+        'description_l10n_*',
```
Comment (chuckharmston) [inline mkt/extensions/indexers.py]: Not sure what the changes to this file have to do with this commit?

File: mkt/extensions/migrations/0015_extension_author.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+from django.db import models, migrations
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('extensions', '0014_soft_delete'),
```
Comment (chuckharmston) [inline mkt/extensions/migrations/0015_extension_author.py]: Is this a real dependency? Seems to me that it depends on nothing but `0001_initial`

File: mkt/extensions/models.py [hunk 7]
```diff
@@ -75,6 +75,7 @@ class Extension(ModelBase):
 
     # Fields for which the manifest is the source of truth - can't be
     # overridden by the API.
+    author = models.CharField(default='', editable=False, max_length=128)
```
Comment (chuckharmston) [inline mkt/extensions/models.py]: Why not nullable?

File: mkt/extensions/tests/test_validation.py [hunk 8]
```diff
@@ -108,16 +118,60 @@ def test_description_missing_valid(self):
...
+
+    def test_author_invalid(self):
+        with self.assertRaises(ParseError):
+            self.validator.validate_author({'author': 42})
+        with self.assertRaises(ParseError):
+            self.validator.validate_author({'author': []})
+        with self.assertRaises(ParseError):
+            self.validator.validate_author({'author': {}})
+        with self.assertRaises(ParseError):
+            self.validator.validate_author({'author': None})
```
Comment (chuckharmston) [inline mkt/extensions/tests/test_validation.py]: `None` is valid here, isn't it?

Comment (chuckharmston) [inline mkt/extensions/tests/test_validation.py]: Imagining a manifest with `{ "author": null }`

**Comment by chuckharmston** [comment]
Seems like there's a lot of drive-by cleanup in here. It'd be nice if we could factor that out into a separate commit, but...

File: mkt/extensions/indexers.py (hunk 5 above)
Comment (diox) [inline mkt/extensions/indexers.py]: drive-by optimization, I forgot to do this earlier

File: mkt/extensions/migrations/0015_extension_author.py (hunk 6 above)
Comment (diox) [inline mkt/extensions/migrations/0015_extension_author.py]: It's how makemigrations generated it - it does that, and I guess it's good to avoid conflicts

File: mkt/extensions/tests/test_validation.py (hunk 8 above)
Comment (diox) [inline mkt/extensions/tests/test_validation.py]: That's precisely what I was imagining and it doesn't feel valid to me.

File: mkt/extensions/models.py (hunk 7 above)
Comment (diox) [inline mkt/extensions/models.py]: I try to avoid it for `CharField`s for new code when I can, because of https://docs.djangoproject.com/en/1.8/ref/models/fields/#null

File: docs/api/topics/firefox_os_addons.rst (hunk 4 above)
Comment (diox) [inline docs/api/topics/firefox_os_addons.rst]: It defaults to an empty string, see below.

File: mkt/extensions/indexers.py (hunk 5 above)
Comment (diox) [inline mkt/extensions/indexers.py]: Splitted into a different commit

**Comment by diox** [comment]
I splitted everything into 3 different commits
